azurerm/internal/services/resource/resource_arm_provider_registration.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/resourceproviders"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmResourceProviderRegistration() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmResourceProviderCreate,
+		Read:   resourceArmResourceProviderRead,
+		Delete: resourceArmResourceProviderDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(30 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Minute),
+			Delete: schema.DefaultTimeout(30 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+		},
+	}
+}
+
+func resourceArmResourceProviderCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Resource.ProvidersClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	resourceProviderNamespace := d.Get("name").(string)
+	expand := ""
+
+	isSkipResourceProviderRegistration := meta.(*clients.Client).Account.SkipResourceProviderRegistration
+
+	if !isSkipResourceProviderRegistration {
+		for resourceProvider := range resourceproviders.Required() {
+			if strings.EqualFold(resourceProviderNamespace, resourceProvider) {
+				return fmt.Errorf(`The Resource Provider %q is automatically registered by Terraform. To manage this Resource Provider Registration with Terraform you need to opt-out of Automatic Resource Provider Registration (by setting 'skip_provider_registration' to 'true' in the Provider block) to avoid conflicting with Terraform`, resourceProviderNamespace)
+			}
+		}
+	}
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, resourceProviderNamespace, expand)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("Error checking for presence of existing Resource Provider Namespace %q: %+v",
+					resourceProviderNamespace, err)
+			}
+		}
+
+		if existing.ID != nil && existing.RegistrationState != nil && *existing.RegistrationState == "Registered" {
+			return tf.ImportAsExistsError("azurerm_resource_provider_registration", *existing.ID)
```

Comment:
Could we maybe have a different error message here, I feel if we display the `ImportAsExistsError` this may confuse end users, WDYT?

Import Error:
```
        Error: A resource with the ID "/subscriptions/*******/providers/Microsoft.PolicyInsights" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test205268533/main.tf line 7:
          (source code not available)
```

Possible New Error:
```
        Error: The "/subscriptions/*******/providers/Microsoft.PolicyInsights" provider is currently already registered in this subscription.
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test205268533/main.tf line 7:
          (source code not available)
```

Reply:
I think the user should still be notified that this needs to imported in to the state. I can change it to 

The "/subscriptions/*******/providers/Microsoft.PolicyInsights" provider is currently already registered in this subscription. To be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_provider_registration" for more information.


@tombuildsstuff @WodansSon - WDYT?

Reply:
@Humoiz , that works for me! 💯

Reply:
Personally, I don't think the message is a blocker, it can be updated in a future release. I'm just hoping we can get this into the current release 🤞

Reply:
I can make a label change very quick but I want @tombuildsstuff to agree with this too. Other than this if there is no other feedback than this is ready to be merged.

Reply:
I want this PR to be merged soon now. This is very minor and I think can be updated outside of this PR.